File: cmdline/ninjakb.js

const fs = require('fs');
const conf = JSON.parse(fs.readFileSync('../conf/conf.json', 'utf8'));
const scancodes = JSON.parse(fs.readFileSync('../frontend/public/scancodes.json', 'utf8'));
const {SerialPort} = require('serialport');

const port = new SerialPort({
  path: conf['device'],
  baudRate: 115200
});

function processKey(key, i, side, layer){
  const n =`0${i}`.slice(-2);
  let t='c';
  let k=null;
  if( scancodes.modifiers.indexOf(key)>=0){
    t='m';
    k=key;
  }else{
    if(key.match(/^M[ULDRB]/)) {
      t='p'
      k=key
    }else{      
      k=scancodes.codes[key];
    } 
    //k=scancodes.codes[key];
  }
  if(k){
    const cmd = `${side}${layer}${n}${t}${k}`;
    port.write(cmd, function(){
      port.flush();
    });
    console.log(cmd);
  }
}
function sendKeys(){
  const { layers } = conf;
  const sides_a=["left","right"]
  const sides_aa=["l","r"]
  for (let l=0; l<layers.length; l++){    
    for (let s=0; s<2; s++){
      const ss=sides_a[s];
      const keys = layers[l][ss];
//      console.log(keys);
      for (let i=0; i<keys.length; i++){
        processKey(keys[i], i, sides_aa[s], l);
      }
    }
  }
}

port.on("open",  function(){
  sendKeys();
  port.write("save");
  console.log('All keys sent');
  console.log('dump0');
  port.write("dump0");
  console.log('dump1');
  port.write("dump1");
  
});

File: cmdline/ninjakb.py
#!/usr/bin/python

import json
import serial
import re
from pathlib import Path

home = str(Path.home())

f = open(home+"/.ninjakb/conf.json")
conf = json.load(f)
f2 = open("../frontend/public/scancodes.json")
scancodes = json.load(f2)

ser = serial.Serial()
ser.baudrate = 115200
ser.port=conf['device']

def sendKeys(keys, layer, side):
	for i, key in enumerate(keys):		
		print("key: "+key)
		t='c';
		k=None;
		if key in scancodes["modifiers"]:
			t='m'
			k=key
		else:
			if re.match('^M[ULDRB]/',key):
				t='p'
				k=key
			else:
				if key in scancodes["codes"]:
					k=scancodes["codes"][key]
		if k:
			cmd = "%s%d%02d%s%s" % (side[0],layer,i,t,k)
			print(cmd)
			ser.write(cmd.encode())

ser.open()
if ser.is_open:
	for l, layer in enumerate(conf['layers']):
		#print(layer)
		sendKeys(layer['left'],l,'l')
		sendKeys(layer['right'],l,'r')

	ser.write(b'save')
	ser.write(b'dump0')
	ser.write(b'dump1')	

f.close